add kahan summation, mean, and variance

gdkrmr opened the pull request:
Interested? There is a problem: in all my tests this was identical to `Sum()` for both `Float32` and `Float64`, I have the suspicion, that the compiler simply optimizes the correction away. Any idea how to prevent this? I already tried `julia --math-mode=ieee -O0`:

```julia
julia> d = (rand(1_000_000) .^ 3) |> sort |> reverse;

julia> d32 = Float32.(d);

julia> 

julia> sum(d) - sum(d32)
-0.012182071630377322

julia> reduce(+, d) - reduce(+, d32)
-0.012182071630377322

julia> foldr(+, d) - foldr(+, d32)
-7.543432063568616

julia> foldl(+, d) - foldl(+, d32)
217.95656791204237

julia> s64 = Sum(Float64); s32 = Sum(Float32);

julia> sum(fit!(s64, d)) - sum(fit!(s32, d))
217.95656791204237

julia> ks64 = KahanSum(Float64); ks32 = KahanSum(Float32);

julia> sum(fit!(ks64, d)) - sum(fit!(ks32, d))
217.95656791204237
```

edit: I am also not sure if the `_merge!` function is correct.

gdkrmr commented:
works fine now!!
```julia
julia> d = (rand(1_000_000) .^ 3) |> sort |> reverse;

julia> d32 = Float32.(d);

julia> sum(d) - sum(d32)
-0.0017536550294607878

julia> reduce(+, d) - reduce(+, d32)
-0.0017536550294607878

julia> foldr(+, d) - foldr(+, d32)
-6.908003657183144

julia> foldl(+, d) - foldl(+, d32)
229.63887133196113

julia> s64 = Sum(Float64); s32 = Sum(Float32);

julia> sum(fit!(s64, d)) - sum(fit!(s32, d))
229.63887133196113

julia> ks64 = KahanSum(Float64); ks32 = KahanSum(Float32);

julia> sum(fit!(ks64, d)) - sum(fit!(ks32, d))
-0.0017536550585646182
```

gdkrmr commented:
There are `KahanMean` and `KahanVariance` now, please test if you get good accuracy:
```julia
using OnlineStats
using Statistics

d = (rand(1_000_000) .^ 3) |> sort |> reverse;
d32 = Float32.(d);
l = length(d)

t = sum(BigFloat.(d))
t - sum(d)
t - reduce(+, d)
t - foldr(+, d)
t - foldl(+, d)
t - sum(fit!(Sum(Float64), d))
t - sum(fit!(KahanSum(Float64), d))

t = sum(BigFloat.(d))
t - sum(d32)
t - reduce(+, d32)
t - foldr(+, d32)
t - foldl(+, d32)
t - sum(fit!(Sum(Float32), d))
t - sum(fit!(KahanSum(Float32), d))


d = (rand(1_000_000) .^ 3);
d32 = Float32.(d);
l = length(d)

t = mean(BigFloat.(d))
t - mean(d)
t - reduce(+, d)/l
t - foldr(+, d)/l
t - foldl(+, d)/l
t - mean(fit!(Mean(), d))
t - mean(fit!(OnlineStats.KahanMean(T = Float64), d))

t = mean(BigFloat.(d32))
t - mean(d32)
t - reduce(+, d32)/l
t - foldr(+, d32)/l
t - foldl(+, d32)/l
t - mean(fit!(OnlineStats.KahanMean(T = Float32), d32))

t = var(BigFloat.(d))
t - var(d)
t - var(fit!(Variance(), d))
t - var(fit!(OnlineStats.KahanVariance(T = Float64), d))

t = var(BigFloat.(d32))
t - var(d32)
t - var(fit!(OnlineStats.KahanVariance(T = Float32), d32))
```

joshday commented:
>  I am not sure if it is legal to simply add the compensations in the _merge! functions.

Me neither...maybe we can pull in someone more familiar with it than myself to weigh in.

> should their accuracy be tested?

I think accuracy tests are pretty important, given that accuracy is the advantage over the normal `Sum`/`Mean`/`Variance`

gdkrmr commented:
`merge!` for `KahanSum` works really well now. `merge!` for `KahanMean` is only as accurate as `Mean` and the unit tests don't pass reliably because of that.

I was thinking the following: As Kahan summation simply saves the lower bits in an extra value, it should be approximately as accurate as doubling precision. What about giving  `Mean` and `Variance` a type parameter, this way Quadruple or BigFloats could be used?

gdkrmr commented:
So what about this PR? The `fit!` functions work pretty well for all, the `merge!` functions only work well for `KahanSum`, which ones do you want to keep?

joshday commented:
> The fit! functions work pretty well for all, the merge! functions only work well for KahanSum

As long as by "pretty well" you mean "correct", I'm happy to add this, but some of the CI failures are from the new tests.   Feel free to comment out the `CovMatrix` tests that use complex numbers, since those are failing randomly 😞.

gdkrmr commented:
Ok great! Under certain circumstances `merge!` for `KahanMean` can be a little bit less accurate than `merge!` for `Mean`. The reason that the tests are failing is because I made them pretty strict I will allow a little bit more inaccuracy. I also still have to add a merge function for `KahanVariance`.

tk3369 commented:
Not that I have any use case for this at the moment, but this is an awesome PR :+1:

gdkrmr commented:
- Tests pass now, the accuracy tests involving `merge!` are pretty brittle for `KahanMean` and `KahanVariance`, I fixed it by setting a random number seed. 
- I added some cross-references and comments to the docs, please check if you are happy with them.
- I also commented out the complex covariance functions.

joshday commented:
This is a solid PR, thanks!  Merging now.
